File: server/app/schema.py
"""旅游方案数据 Schema

与前端 app/js/schema.js 保持字段一致。LLM 返回 JSON 必须满足本结构。
所有可选字段携带 source_url + source_type，用于来源追溯与动态优化。
"""
from __future__ import annotations
from dataclasses import dataclass, field, asdict
from typing import Optional, List, Dict, Any


# 顶层必需字段（与 schema.js REQUIRED_FIELDS 对齐）
REQUIRED_TOP_FIELDS = {
    "overview",
    "route",
    "daily",
    "attractions",
    "accommodation",
    "food",
    "budget",
    "tips",
}

OVERVIEW_FIELDS = [
    "title", "dates", "days", "travelers", "departure",
    "route_theme", "highlights", "mileage", "suitable_people", "tips",
]

BUDGET_FIELDS = ["accommodation", "transport", "food", "tickets", "other", "total", "per_person"]


def validate_plan(plan: Dict[str, Any]) -> Dict[str, Any]:
    """返回 {valid: bool, errors: [str], warnings: [str]}。"""
    errors: List[str] = []
    warnings: List[str] = []
    if not isinstance(plan, dict):
        return {"valid": False, "errors": ["根节点不是对象"], "warnings": []}

    for f in REQUIRED_TOP_FIELDS:
        if f not in plan:
            errors.append(f"缺少顶层字段: {f}")

    ov = plan.get("overview")
    if isinstance(ov, dict):
        for sf in OVERVIEW_FIELDS:
            if sf not in ov:
                warnings.append(f"overview.{sf} 缺失")
    else:
        errors.append("overview 不是对象")

    bg = plan.get("budget")
    if isinstance(bg, dict):
        for sf in BUDGET_FIELDS:
            if sf not in bg:
                warnings.append(f"budget.{sf} 缺失")
    else:
        errors.append("budget 不是对象")

    if isinstance(plan.get("daily"), list) and not plan["daily"]:
        errors.append("daily 为空数组")
    if isinstance(plan.get("attractions"), list) and not plan["attractions"]:
        errors.append("attractions 为空数组")

    # 来源追溯：至少 70% 条目带 source_url（仅做 warning，不阻塞）
    src_coverage = _source_coverage(plan)
    if src_coverage["total"] > 0 and src_coverage["with_url"] / src_coverage["total"] < 0.7:
        warnings.append(
            f"source_url 覆盖率 {src_coverage['with_url']}/{src_coverage['total']} < 70%"
        )

    return {"valid": len(errors) == 0, "errors": errors, "warnings": warnings}


def _source_coverage(plan: Dict[str, Any]) -> Dict[str, int]:
    total = 0
    with_url = 0
    for key in ("attractions", "food", "accommodation"):
        items = plan.get(key, [])
        if isinstance(items, list):
            for it in items:
                if isinstance(it, dict):
                    total += 1
                    if it.get("source_url"):
                        with_url += 1
    for d in plan.get("daily", []) or []:
        if not isinstance(d, dict):
            continue
        for key in ("attractions", "food"):
            for it in d.get(key, []) or []:
                if isinstance(it, dict):
                    total += 1
                    if it.get("source_url"):
                        with_url += 1
    return {"total": total, "with_url": with_url}


def make_response(
    plan: Dict[str, Any],
    sources: List[Dict[str, str]],
    warnings: List[Dict[str, str]],
    session_id: str,
    version_id: Optional[str] = None,
) -> Dict[str, Any]:
    """组装 /generate 标准响应。"""
    return {
        "plan": plan,
        "sources": sources,
        "warnings": warnings,
        "session_id": session_id,
        "version_id": version_id or session_id,
        "source_coverage": _source_coverage(plan),
    }

File: app/js/schema.js
// 旅游方案数据 Schema 与校验
// 定义 LLM 应返回的 JSON 结构，并提供校验函数

/**
 * @typedef {Object} Overview
 * @property {string} title - 行程标题
 * @property {string} dates - 出行日期描述
 * @property {number} days - 天数
 * @property {number} travelers - 出行人数
 * @property {string} departure - 出发城市
 * @property {string} route_theme - 路线主题
 * @property {string[]} highlights - 核心亮点
 * @property {string} mileage - 全程里程
 * @property {string} suitable_people - 适合人群
 * @property {string} tips - 温馨提示
 */

/**
 * @typedef {Object} RouteNode
 * @property {string} name - 城市/节点名
 * @property {string} day - 对应天数，如 "Day 1"
 */

/**
 * @typedef {Object} DailyPlan
 * @property {number} day - 第几天
 * @property {string} date - 日期
 * @property {string} city - 城市
 * @property {string} title - 当日标题
 * @property {string} subtitle - 副标题
 * @property {Attraction[]} attractions - 当日景点
 * @property {Food[]} food - 当日美食
 * @property {string} transport - 交通说明
 * @property {string} accommodation - 住宿说明
 */

/**
 * @typedef {Object} Attraction
 * @property {string} name - 景点名
 * @property {string} desc - 描述
 * @property {string} duration - 建议游玩时长
 * @property {string} ticket - 门票
 * @property {string[]} tags - 标签
 * @property {string} image_query - 图片搜索词
 */

/**
 * @typedef {Object} Accommodation
 * @property {string} name - 住宿名称
 * @property {string} area - 区域
 * @property {string} price_range - 价格区间
 * @property {string} desc - 描述
 */

/**
 * @typedef {Object} Food
 * @property {string} name - 美食名
 * @property {string} desc - 描述
 * @property {string} image_query - 图片搜索词
 */

/**
 * @typedef {Object} Budget
 * @property {number} accommodation - 住宿
 * @property {number} transport - 交通
 * @property {number} food - 餐饮
 * @property {number} tickets - 门票
 * @property {number} other - 其他
 * @property {number} total - 合计
 * @property {number} per_person - 人均
 */

/**
 * @typedef {Object} TravelPlan
 * @property {Overview} overview
 * @property {RouteNode[]} route
 * @property {DailyPlan[]} daily
 * @property {Attraction[]} attractions
 * @property {Accommodation[]} accommodation
 * @property {Food[]} food
 * @property {Budget} budget
 * @property {string[]} tips
 */

const REQUIRED_FIELDS = {
  overview: ['title', 'dates', 'days', 'travelers', 'departure', 'route_theme', 'highlights', 'mileage', 'suitable_people', 'tips'],
  route: null,
  daily: null,
  attractions: null,
  accommodation: null,
  food: null,
  budget: ['accommodation', 'transport', 'food', 'tickets', 'other', 'total', 'per_person'],
  tips: null
};

/**
 * 校验旅游方案 JSON
 * @param {any} json
 * @returns {{valid: boolean, errors: string[]}}
 */
export function validatePlan(json) {
  const errors = [];
  if (!json || typeof json !== 'object') {
    return { valid: false, errors: ['根节点不是对象'] };
  }

  for (const [field, subFields] of Object.entries(REQUIRED_FIELDS)) {
    if (!(field in json)) {
      errors.push(`缺少顶层字段: ${field}`);
      continue;
    }
    if (subFields && typeof json[field] === 'object' && !Array.isArray(json[field])) {
      for (const sf of subFields) {
        if (!(sf in json[field])) {
          errors.push(`overview.${sf} 或 budget.${sf} 缺失: ${field}.${sf}`);
        }
      }
    }
  }

  // 检查 daily 至少有一项
  if (Array.isArray(json.daily) && json.daily.length === 0) {
    errors.push('daily 为空数组');
  }
  // 检查 attractions 至少有一项
  if (Array.isArray(json.attractions) && json.attractions.length === 0) {
    errors.push('attractions 为空数组');
  }

  return { valid: errors.length === 0, errors };
}

/**
 * 从目的地+日期生成稳定 planId
 */
export function makePlanId(dest, days) {
  const base = `${dest}-${days}-${new Date().toISOString().slice(0, 10)}`;
  let hash = 0;
  for (let i = 0; i < base.length; i++) {
    hash = ((hash << 5) - hash) + base.charCodeAt(i);
    hash |= 0;
  }
  return `plan_${Math.abs(hash)}`;
}
